cmd/manager/main.go

```diff
@@ -0,0 +1,187 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"os"
+	"runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	// Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+
+	"github.com/bluek8s/kubedirector/pkg/apis"
+	"github.com/bluek8s/kubedirector/pkg/controller"
+
+	kdv1 "github.com/bluek8s/kubedirector/pkg/apis/kubedirector.bluedata.io/v1alpha1"
+	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
+	"github.com/operator-framework/operator-sdk/pkg/leader"
+	"github.com/operator-framework/operator-sdk/pkg/log/zap"
+	"github.com/operator-framework/operator-sdk/pkg/metrics"
+	"github.com/operator-framework/operator-sdk/pkg/restmapper"
+	sdkVersion "github.com/operator-framework/operator-sdk/version"
+	"github.com/spf13/pflag"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
+)
+
+// Change below variables to serve metrics on different host or port.
+var (
+	metricsHost       = "0.0.0.0"
+	metricsPort int32 = 8383
+)
+var log = logf.Log.WithName("cmd")
+
+func printVersion() {
+	log.Info(fmt.Sprintf("Go Version: %s", runtime.Version()))
+	log.Info(fmt.Sprintf("Go OS/Arch: %s/%s", runtime.GOOS, runtime.GOARCH))
+	log.Info(fmt.Sprintf("Version of operator-sdk: %v", sdkVersion.Version))
+}
+
+func main() {
+	// TODO: the operator-sdk switched to a new log implementation.
+	//       should we migrate the shared logger code?
+
+	// Add the zap logger flag set to the CLI. The flag set must
+	// be added before calling pflag.Parse().
+	pflag.CommandLine.AddFlagSet(zap.FlagSet())
+
+	// Add flags registered by imported packages (e.g. glog and
+	// controller-runtime)
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+
+	pflag.Parse()
+
+	// Use a zap logr.Logger implementation. If none of the zap
+	// flags are configured (or if the zap flag set is not being
+	// used), this defaults to a production zap logger.
+	//
+	// The logger instantiated here can be changed to any logger
+	// implementing the logr.Logger interface. This logger will
+	// be propagated through the whole operator, generating
+	// uniform and structured logs.
+	logf.SetLogger(zap.Logger())
+
+	printVersion()
+
+	namespace, err := k8sutil.GetWatchNamespace()
+	if err != nil {
+		log.Error(err, "Failed to get watch namespace")
+		os.Exit(1)
+	}
+
+	// Get a config to talk to the apiserver
+	cfg, err := config.GetConfig()
+	if err != nil {
+		log.Error(err, "")
+		os.Exit(1)
+	}
+
+	ctx := context.TODO()
+
+	// Become the leader before proceeding
+	err = leader.Become(ctx, "kubedirector-lock")
+	if err != nil {
+		log.Error(err, "")
+		os.Exit(1)
+	}
+
+	// Create a new Cmd to provide shared dependencies and start components
+	mgr, err := manager.New(cfg, manager.Options{
+		Namespace:          namespace,
```

Comment:
So in current KD, when we watch for KubeDirectorCluster we set namespace to emptystring so that we end up watching all namespaces. For KubeDirectorConfig on the other hand we only watch in a single namespace. it's this code in the old main.go:
```go
	watchParams := []watchInfo{
		{
			kind: "KubeDirectorCluster",
			// The resync period essentially determines how granularly we can detect
			// the completion of cluster config changes. Making this too small can
			// actually be bad in that there is benefit to batch-resolving changes,
			// within KubeDirector but also especially with the cluster's app config
			// scripts.
			resyncPeriod: time.Duration(30) * time.Second,
			namespace:    "",
		},
		{
			kind:         "KubeDirectorConfig",
			resyncPeriod: 0,
			namespace:    namespace,
		},
	}
```

I'm not 100% sure about the behavior of this new setup -- what the exact implications are of passing this namespace to the manager -- but in any case we want to preserve the old behavior where we can watch for KubeDirectorCluster objects appearing in any namespace.

Reply:
c, err := controller.New("kubedirectorcluster-controller", mgr, controller.Options{Reconciler: r})
	if err != nil {
		return err
	}

	// Watch for changes to primary resource KubeDirectorCluster
	err = c.Watch(&source.Kind{Type: &v1alpha1.KubeDirectorCluster{}}, &handler.EnqueueRequestForObject{})
	if err != nil {
		return err
	}

When we create the controller and configure the watch we don't specify a namespace so I'm assuming we watch for objects in all namespaces based on their Group/Version/Kind.

Reply:
Hmm. Do you know what the namespace given to the manager is used for then?

Anyway this will be easy enough to test/verify.